storage_file/models/storage_backend.py

```diff
@@ -0,0 +1,171 @@
+            elif "." in param:
+                backend = env.ref(param, raise_if_not_found=False)
+                if backend:
+                    backend_id = backend.id
+        if not backend_id and default_fallback:
+            backend = env.ref(cls._default_backend_xid, raise_if_not_found=False)
+            if backend:
+                backend_id = backend.id
+            else:
+                _logger.warn("No backend found, no default fallback found.")
+        return backend_id
+
+    @api.depends(
+        "served_by",
+        "base_url",
+        "directory_path",
+        "url_include_directory_path",
+    )
+    def _compute_base_url_for_files(self):
+        for record in self:
+            record.base_url_for_files = record._get_base_url_for_files()
+
+    def _get_base_url_for_files(self):
+        """Retrieve base URL for files."""
+        backend = self.sudo()
+        parts = []
+        if backend.served_by == "external":
+            parts = [backend.base_url or ""]
+            if backend.url_include_directory_path and backend.directory_path:
+                parts.append(backend.directory_path)
+        return "/".join(parts)
+
+    def action_recompute_base_url_for_files(self):
+        """Refresh base URL for files.
+
+        Rationale: all the params for computing this URL might come from server env.
+        When this is the case, the URL - being stored - might be out of date.
+        This is because changes to server env fields are not detected at startup.
+        Hence, let's offer an easy way to promptly force this manually when needed.
+        """
+        self._compute_base_url_for_files()
+        self.env["storage.file"].invalidate_model(["url"])
+
+    def _get_base_url_from_param(self):
+        base_url_param = (
+            "report.url" if self.env.context.get("print_report_pdf") else "web.base.url"
+        )
+        return self.env["ir.config_parameter"].sudo().get_param(base_url_param)
+
+    def _get_url_for_file(self, storage_file, exclude_base_url=False):
+        """Return final full URL for given file."""
+        backend = self.sudo()
+        if backend.served_by == "odoo":
+            parts = [
+                self._get_base_url_from_param() if not exclude_base_url else "/",
+                "storage.file",
+                storage_file.slug,
+            ]
+        else:
+            base_url = backend.base_url_for_files
+            if exclude_base_url:
+                base_url = base_url.replace(backend.base_url, "") or "/"
+            parts = [base_url, storage_file.relative_path or ""]
+        return "/".join([x.rstrip("/") for x in parts if x])
+
+    def _register_hook(self):
+        super()._register_hook()
+        backends = self.search([]).filtered(
+            lambda x: x._get_base_url_for_files() != x.base_url_for_files
+        )
+        if not backends:
+            return
+        sql = SQL(f"SELECT id FROM {self._table} WHERE ID IN %s FOR UPDATE")
```

Comment:
If you are applying `SQL` here because it protects better against SQL injection (see comment below), then you need to pass the table name as an `SQL.identifier`.